### PR TITLE
En 6314/software version to block

### PR DESCRIPTION
**Pull request by sasurobert**
Added software version to the block header and check for it to be non-zero length and less than 10.

### CODE REVIEW

File: data/block/block.go [hunk 1]
```diff
@@ -181,6 +182,14 @@ func (h *Header) CheckChainID(reference []byte) error {
 	return nil
 }
 
+// CheckSoftwareVersion returns nil if the software version has the correct length
+func (h *Header) CheckSoftwareVersion() error {
+	if len(h.SoftwareVersion) == 0 && len(h.SoftwareVersion) > core.MaxSoftwareVersionLength {
```
Comment (SebastianMarian) [inline data/block/block.go:187]: || instead &&

File: core/constants.go [hunk 2]
```diff
@@ -386,3 +386,6 @@ const SleepTimeBetweenCreateDBRetries = 5 * time.Second
 
 // ElrondProtectedKeyPrefix is the key prefix which is protected from writing in the trie - only for special builtin functions
 const ElrondProtectedKeyPrefix = "ELROND"
+
+// MaxSoftwareVersionLength represents the maximum length for the software version to be saved in block header
+const MaxSoftwareVersionLength = 10
```
Comment (SebastianMarian) [inline core/constants.go:391]: MaxSoftwareVersionLengthInBytes?

File: data/block/metaBlock.go [hunk 3]
```diff
@@ -161,6 +161,14 @@ func (m *MetaBlock) CheckChainID(reference []byte) error {
 	return nil
 }
 
+// CheckSoftwareVersion returns nil if the software version has the correct length
+func (m *MetaBlock) CheckSoftwareVersion() error {
+	if len(m.SoftwareVersion) == 0 && len(m.SoftwareVersion) > core.MaxSoftwareVersionLength {
```
Comment (SebastianMarian) [inline data/block/metaBlock.go:166]: || instead &&

File: data/block/block.go [hunk 4]
```diff
@@ -181,6 +182,14 @@ func (h *Header) CheckChainID(reference []byte) error {
 	return nil
 }
 
+// CheckSoftwareVersion returns nil if the software version has the correct length
+func (h *Header) CheckSoftwareVersion() error {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply data/block/block.go:186]: added TODO.

File: process/block/metablock.go [hunk 5]
```diff
@@ -105,6 +106,7 @@ func NewMetaProcessor(arguments ArgMetaProcessor) (*metaProcessor, error) {
 		blockChain:             arguments.BlockChain,
 		stateCheckpointModulus: arguments.StateCheckpointModulus,
 		genesisNonce:           genesisHdr.GetNonce(),
+		version:                arguments.Version[:sftVersionLength],
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go:109]: done

File: process/block/metablock.go [hunk 6]
```diff
@@ -80,6 +80,7 @@ func NewMetaProcessor(arguments ArgMetaProcessor) (*metaProcessor, error) {
 		return nil, process.ErrNilValidatorStatistics
 	}
 
+	sftVersionLength := core.MinInt(core.MaxSoftwareVersionLength, len(arguments.Version))
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go:83]: done.

File: update/errors.go [hunk 7]
```diff
@@ -196,3 +196,9 @@ var ErrNilHardForkBlockProcessor = errors.New("nil hard fork block processor")
 
 // ErrNilTrieStorageManager signals that nil trie storage manager has been provided
 var ErrNilTrieStorageManager = errors.New("nil trie storage")
+
+// ErrEmptySoftwareVersion signals that empty software version was provided
+var ErrEmptySoftwareVersion = errors.New("empty software version")
+
+// ErrEmptyChainID signals that empty chain ID was provided
+var ErrEmptyChainID = errors.New("empty software version")
```
Comment (SebastianMarian) [inline update/errors.go:201]: empty chain ID

File: update/process/afterHardFork.go [hunk 8]
```diff
@@ -80,7 +80,7 @@ func (a *afterHardFork) CreateAllBlocksAfterHardfork(
 			return nil, nil, update.ErrNilHardForkBlockProcessor
 		}
 
-		hdr, body, err := blockProcessor.CreateNewBlock(chainID, round, nonce, epoch)
+		hdr, body, err := blockProcessor.CreateNewBlock(chainID, "", round, nonce, epoch)
```
Comment (SebastianMarian) [inline update/process/afterHardFork.go]: "" will be replaced later with a real version ?

File: update/process/shardBlock.go [hunk 9]
```diff
@@ -81,12 +89,14 @@ func (s *shardBlockCreator) CreateNewBlock(
 		return nil, nil, err
 	}
 
+	sftVersionLength := core.MinInt(core.MaxSoftwareVersionLength, len(version))
```
Comment (SebastianMarian) [inline update/process/shardBlock.go]: You can add in core package a method like "TrimSoftwareVersion(version) trimmedVersion" and use it everywhere you needed

File: update/process/metaBlock.go [hunk 10]
```diff
@@ -82,6 +90,7 @@ func (m *metaBlockCreator) CreateNewBlock(
 	blockBody := &block.Body{
 		MiniBlocks: make([]*block.MiniBlock, 0),
 	}
+	sftVersionLength := core.MinInt(core.MaxSoftwareVersionLength, len(version))
```
Comment (SebastianMarian) [inline update/process/metaBlock.go]: You can add in core package a method like "TrimSoftwareVersion(version) trimmedVersion" and use it everywhere you needed

File: process/block/shardblock.go [hunk 11]
```diff
@@ -53,6 +53,7 @@ func NewShardProcessor(arguments ArgShardProcessor) (*shardProcessor, error) {
 		return nil, process.ErrNilTransactionPool
 	}
 
+	sftVersionLength := core.MinInt(core.MaxSoftwareVersionLength, len(arguments.Version))
```
Comment (SebastianMarian) [inline process/block/shardblock.go]: You can add in core package a method like "TrimSoftwareVersion(version) trimmedVersion" and use it everywhere you needed

File: process/block/metablock.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/block/metablock.go]: You can add in core package a method like "TrimSoftwareVersion(version) trimmedVersion" and use it everywhere you needed

File: data/block/block.go (hunk 1 above)
Comment (sasurobert) [reply data/block/block.go:187]: done

File: core/constants.go (hunk 2 above)
Comment (sasurobert) [reply core/constants.go:391]: done

File: data/block/metaBlock.go (hunk 3 above)
Comment (sasurobert) [reply data/block/metaBlock.go:166]: done

File: update/errors.go (hunk 7 above)
Comment (sasurobert) [reply update/errors.go:201]: done

File: update/process/afterHardFork.go (hunk 8 above)
Comment (sasurobert) [reply update/process/afterHardFork.go]: refactored

File: update/process/shardBlock.go (hunk 9 above)
Comment (sasurobert) [reply update/process/shardBlock.go]: done

File: update/process/metaBlock.go (hunk 10 above)
Comment (sasurobert) [reply update/process/metaBlock.go]: done

File: process/block/shardblock.go (hunk 11 above)
Comment (sasurobert) [reply process/block/shardblock.go]: done

**Pull request merged**